### PR TITLE
IMB-37 & IMB-43: Added Exception and Removal Conditions 3 & 4 pages

### PR DESCRIPTION
**Pull request by mislam987**
## What
- Added Exception page - [IMB-37](https://collaboration.homeoffice.gov.uk/jira/browse/IMB-37)
- Added Removal Conditions 3 & 4 pages - [IMB-43](https://collaboration.homeoffice.gov.uk/jira/browse/IMB-43)

## Why
- Added the new screens to match the designs on Figma, as part of the flow for the IMA MVP

## How
- Added steps, fields and templates to the index.js file in the /ima folder
- Added corresponding field configurations in the fields.json, pages.json, and index.js (/fields folder) files
- Added new html pages and partials, and referred to them in the /fields index.js file to enable the toggling of text areas where applicable
- Added exception page details to the summary page. However, left out the removal conditions 3 & 4 pages from the summary page because it is included within [IMB-53](https://github.com/UKHomeOffice/ima/tree/IMB-53)

## Testing
- Tests passing locally

### CODE REVIEW

File: apps/ima/index.js [hunk 1]
```diff
@@ -33,9 +33,38 @@ module.exports = {
       behaviours: SaveFormSession,
       fields: ['who-are-you'],
       locals: { showSaveAndExit: true },
-      next: '/confirm', // TO BE UPDATED AS STEPS ARE ADDED
+      next: '/exception',
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/index.js:36]: can you keep the comment that this is to be updated as steps are added

File: apps/ima/views/permission.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{{<partials-page}}
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/views/permission.html:1]: Do you need a html template for this page? Wouldn't it be created automatically as it is the file name same as the step name?

File: apps/ima/views/permission-to-enter-or-stay.html [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{{<partials-page}}
+  {{$page-content}}
+    {{#renderField}}permission-to-enter-or-stay{{/renderField}}
+    {{#input-submit}}continue{{/input-submit}}
+  {{/page-content}}
+{{/partials-page}}
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/views/permission-to-enter-or-stay.html:6]: Is this file needed?

File: apps/ima/index.js [hunk 4]
```diff
@@ -33,9 +33,38 @@ module.exports = {
...
     },
+    '/exception': {
+      behaviours: SaveFormSession,
+      fields: [
+        'does-exception-apply',
+        'does-exception-apply-detail'
+      ],
+      template: 'does-exception-apply',
+      locals: { showSaveAndExit: true },
+      next: '/threatened-life-or-liberty'
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/index.js:47]: Can you add a comment that this will need to be updated once the correct next step is added

**Review by Rhodine-orleans-lindsay — CHANGES_REQUESTED** [review]
A few changes needed (see comments). Can you also add your fields to the summary-data-sections file so that you can see them on the summary page at the beginning of the form and on the confirm page? Can you also make your commit message more specific to your changes according to the team format (as this is the only commit going in you don't need to say that you moved it from your previous branch, just write the commit as if it is your original branch)?

File: apps/ima/views/permission-to-enter-or-stay.html (hunk 3 above)
Comment (mislam987) [reply apps/ima/views/permission-to-enter-or-stay.html:6]: Removed this file.

File: apps/ima/index.js (hunk 4 above)
Comment (mislam987) [reply apps/ima/index.js:47]: Added in the comment for this step.

File: apps/ima/index.js (hunk 1 above)
Comment (mislam987) [reply apps/ima/index.js:36]: Added the comment to this step.

File: apps/ima/index.js [hunk 5]
```diff
@@ -33,9 +33,40 @@ module.exports = {
...
+    },
+    '/threatened-life-or-liberty': {
+      behaviours: SaveFormSession,
+      fields: [
+        'is-life-threatened',
+        'life-threatened-detail'
+      ],
+      template: 'life-or-liberty-threatened',
+      locals: { showSaveAndExit: true },
+      continueOnEdit: true, // TO BE UPDATED AS STEPS ARE ADDED
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/index.js:57]: why is this comment here?

File: apps/ima/index.js [hunk 6]
```diff
@@ -33,9 +33,40 @@ module.exports = {
...
     },
+    '/exception': {
+      behaviours: SaveFormSession,
+      fields: [
+        'does-exception-apply',
+        'does-exception-apply-detail'
+      ],
+      template: 'does-exception-apply',
+      locals: { showSaveAndExit: true },
+      next: '/threatened-life-or-liberty' // TO BE UPDATED AS STEPS ARE ADDED
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/index.js:47]: I don't think the comment is needed here as '/threatened-life-or-liberty'  is the correct next step for this step

File: apps/ima/sections/summary-data-sections.js [hunk 7]
```diff
@@ -17,5 +17,20 @@ module.exports = {
...
+        step: '/exception',
+        field: 'does-exception-apply',
+        parse: (list, req) => {
+          if (!req.sessionModel.get('steps').includes('/exception')) {
+            return null;
+          }
+          return req.sessionModel.get('does-exception-apply') === 'yes' ?
+            'Yes' + '\nDetails added' : 'No';
+        }
+      } // Removal conditions 3 & 4 are included in the summary section as part of IMB-53, so have not been added here
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/sections/summary-data-sections.js:33]: Since that section will be added as part of another ticket you don't need to add this comment

File: apps/ima/translations/src/en/fields.json [hunk 8]
```diff
@@ -17,5 +17,47 @@
         "hint": "You cannot give immigration advice. You must answer all the questions as if you are the person named on the duty to remove letter."
       }
     }
+  },
+  "does-exception-apply":{
+    "legend": " ",
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/translations/src/en/fields.json:22]: why is this empty? You can use` legend: {
className: 'visually hidden'}` if you need to instead of making it an empty string

File: apps/ima/translations/src/en/fields.json [hunk 9]
```diff
@@ -17,5 +17,47 @@
...
+      "no": {
+        "label": "No"
+      }
+    }
+  },
+  "does-exception-apply-detail": {
+    "label": "Explain why"
+  },
+  "is-life-threatened": {
+    "legend": " ",
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/translations/src/en/fields.json:36]: why is this empty? See previous comment on this

File: apps/ima/translations/src/en/fields.json [hunk 10]
```diff
@@ -17,5 +17,47 @@
...
+      "yes": {
+        "label": "Yes"
+      },
+      "no": {
+        "label": "No"
+      }
+    }
+  },
+  "life-threatened-detail": {
+    "label": " "
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/translations/src/en/fields.json:47]: why is this empty? You can use` labelClassName: 'visually hidden'` if you need to instead of making it an empty string

File: apps/ima/views/does-exception-apply.html [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{{<partials-page}}
+  {{$page-content}}
+    <a href="">Read the guidance on exceptions</a> to the duty to remove.
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/views/does-exception-apply.html:3]: do we have the link for this?

File: apps/ima/translations/src/en/pages.json [hunk 12]
```diff
@@ -21,12 +21,25 @@
   "token-invalid": {
     "header": "This link has expired"
   },
+  "exception": {
+    "header": "Does an exception to the duty to remove apply to you?"
+  },
+  "threatened-life-or-liberty": {
+    "header": "Did you come directly to the UK from a country where your life or liberty was threatened?",
+    "p1":"Select 'No' if you passed through a country where your life or liberty was not threatened."
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/translations/src/en/pages.json:29]: it looks like this text has been added to the markdown file - is this still needed?

File: apps/ima/views/content/life-or-liberty-threatened.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<span class="threatened-paragraph">Your life of liberty must have been threatened because of your:</span>
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/views/content/life-or-liberty-threatened.md:1]: what is the purpose of this class? There doesn't seem to be any styling for it in the css file. If it is not needed you can remove the span tag completely.

File: apps/ima/fields/index.js [hunk 14]
```diff
@@ -6,5 +6,79 @@ module.exports = {
...
+    validate: 'required'
+  },
+  'does-exception-apply-detail': {
+    labelClassName: 'bold',
+    mixin: 'textarea',
+    attributes: [{
+      attribute: 'rows',
+      value: 5
+    }],
+    validate: ['required', 'notUrl', { type: 'regex', arguments: /^[^\[\]\|<>]{1,15000}$/ }],
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/fields/index.js:29]: I think you discussed the separating the maxlength and regex validations with Sulthan? Can you make those changes here as well?

File: apps/ima/fields/index.js [hunk 15]
```diff
@@ -6,5 +6,79 @@ module.exports = {
...
+    }],
+    validate: 'required'
+  },
+  'life-threatened-detail': {
+    mixin: 'textarea',
+    attributes: [{
+      attribute: 'rows',
+      value: 5
+    }],
+    validate: ['required', 'notUrl', { type: 'regex', arguments: /^[^\[\]\|<>]{1,15000}$/ }],
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/fields/index.js:53]: separate the maxlength and regex validation as mentioned above

File: apps/ima/fields/index.js [hunk 16]
```diff
@@ -6,5 +6,79 @@ module.exports = {
...
+    validate: 'required'
+  },
+  'permission-to-enter-or-stay-detail': {
+    labelClassName: 'bold',
+    mixin: 'textarea',
+    attributes: [{
+      attribute: 'rows',
+      value: 5
+    }],
+    validate: ['required', 'notUrl', { type: 'regex', arguments: /^[^\[\]\|<>]{1,15000}$/ }],
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/fields/index.js:78]: separate the maxlength and regex validation as mentioned above

File: apps/ima/translations/src/en/fields.json [hunk 17]
```diff
@@ -17,5 +17,47 @@
...
+  },
+  "permission-to-enter-or-stay": {
+    "legend": " ",
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/translations/src/en/fields.json:50]: why is this empty? If is you make `isPageHeading` true for this field, you can set the legend here and remove the header from pages, json.

File: apps/ima/fields/index.js [hunk 18]
```diff
@@ -6,5 +6,79 @@ module.exports = {
...
+    dependent: {
+      field: 'is-life-threatened',
+      value: 'yes'
+    }
+  },
+  'permission-to-enter-or-stay': {
+    isPageHeading: false,
```
Comment (Rhodine-orleans-lindsay) [inline apps/ima/fields/index.js:60]: Why is this false?

**Review by Rhodine-orleans-lindsay — CHANGES_REQUESTED** [review]
Some suggestions and questions(see comments)

File: apps/ima/index.js (hunk 6 above)
Comment (mislam987) [reply apps/ima/index.js:47]: According to the Figma design, the next step would be the 'removal-condition' screen (IMB-53)

File: apps/ima/views/does-exception-apply.html (hunk 11 above)
Comment (mislam987) [reply apps/ima/views/does-exception-apply.html:3]: We don't have the link yet.

**Review by Rhodine-orleans-lindsay — COMMENTED** [review]
Can you also squash and write your commits according to the recommended format i.e.

> JIRA NUMBER ticket title
> -  bullet point 1
> - bullet point 2, etc

**Review by Rhodine-orleans-lindsay — APPROVED** [review]
Great work 🙂
